src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -300,6 +284,61 @@ UpdateRelationsToLocalShardTables(Node *node, List *relationShardList)
 								 relationShardList, QTW_EXAMINE_RTES_BEFORE);
 	}
 
+	if (IsA(node, OnConflictExpr))
+	{
+		OnConflictExpr *onConflict = (OnConflictExpr *) node;
+		Oid constraintId = onConflict->constraint;
+
+		if (!OidIsValid(constraintId))
+		{
+			return false;
+		}
+
+		Oid constraintRelationId = InvalidOid;
```

Comment:
Can we put this block to a method, this block is replacing `relationConstraintId` with `shardConstraintId`. So we can do the following:

```c
/* ReplaceRelationConstraintByShardConstraint replaces the relation constraint with its corresponding 
 * shard constraint.
 */
static bool ReplaceRelationConstraintByShardConstraint(Oid constraintId, List* relationShardList) {
		HeapTuple heapTuple = SearchSysCache1(CONSTROID, ObjectIdGetDatum(constraintId));
		if (HeapTupleIsValid(heapTuple))
		{
			Form_pg_constraint contup = (Form_pg_constraint) GETSTRUCT(heapTuple);

			constraintRelationId = contup->conrelid;
			ReleaseSysCache(heapTuple);
		}

		/*
		 * We can return here without calling the walker function, since we know there
		 * will be no possible tables or constraints after this point, by the syntax.
		 */
		if (!OidIsValid(constraintRelationId))
		{
			ereport(ERROR, (errmsg("Invalid relation id (%u) for constraint: %s",
								   constraintRelationId, get_constraint_name(
									   constraintId))));
			return false;
		}

		RelationShard *relationShard = FindRelationShard(constraintRelationId,
														 relationShardList);

		if (relationShard != NULL)
		{
			char *constraintName = get_constraint_name(constraintId);

			AppendShardIdToName(&constraintName, relationShard->shardId);

			Oid shardOid = GetTableLocalShardOid(relationShard->relationId,
												 relationShard->shardId);

			Oid shardConstraintId = get_relation_constraint_oid(shardOid,
																constraintName,
																false);

			onConflict->constraint = shardConstraintId;

			return false;
		}
        return true;
}
```

In `UpdateRelationsToLocalShardTables` you can do:

```c
	if (IsA(node, OnConflictExpr))
	{
		OnConflictExpr *onConflict = (OnConflictExpr *) node;
		Oid constraintId = onConflict->constraint;
        bool stopWalk =  ReplaceRelationConstraintByShardConstraint(constraintId, relationShardList);
        return stopWalk;
}
```